src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs

```diff
@@ -163,33 +163,25 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, Action<Ht
     {
         ArgumentNullException.ThrowIfNull(configureOptions);
 
-        var options = new HttpsConnectionAdapterOptions();
-        listenOptions.KestrelServerOptions.ApplyHttpsDefaults(options);
-        configureOptions(options);
-        listenOptions.KestrelServerOptions.ApplyDefaultCert(options);
-
-        if (options.ServerCertificate == null && options.ServerCertificateSelector == null)
+        return listenOptions.UseHttps(new Lazy<HttpsConnectionAdapterOptions>(() =>
         {
-            throw new InvalidOperationException(CoreStrings.NoCertSpecifiedNoDevelopmentCertificateFound);
-        }
+            // We defer configuration of the https options until build time so that the IConfiguration will be available.
+            // This is particularly important in docker containers, where the docker tools use IConfiguration to tell
+            // us where the development certificates have been mounted.
 
-        return listenOptions.UseHttps(options);
-    }
-
-    // Use Https if a default cert is available
-    internal static bool TryUseHttps(this ListenOptions listenOptions)
-    {
-        var options = new HttpsConnectionAdapterOptions();
-        listenOptions.KestrelServerOptions.ApplyHttpsDefaults(options);
-        listenOptions.KestrelServerOptions.ApplyDefaultCert(options);
+            var options = new HttpsConnectionAdapterOptions();
+            listenOptions.KestrelServerOptions.ApplyHttpsDefaults(options);
+            configureOptions(options);
+            listenOptions.KestrelServerOptions.ApplyDefaultCert(options);
+            options.HttpProtocols = listenOptions.Protocols;
 
-        if (options.ServerCertificate == null && options.ServerCertificateSelector == null)
-        {
-            return false;
-        }
+            if (options.ServerCertificate == null && options.ServerCertificateSelector == null)
+            {
+                throw new InvalidOperationException(CoreStrings.NoCertSpecifiedNoDevelopmentCertificateFound);
+            }
```

Comment:
What's the old and new call stack for this exception? I worry that we're reducing visibility/immediacy of this exception to the caller so it will be harder to debug.

Can we at least run lines the first few lines before the lazy so we don't change their behavior/lifetime?
```
            var options = new HttpsConnectionAdapterOptions();
            listenOptions.KestrelServerOptions.ApplyHttpsDefaults(options);
            configureOptions(options);
            options.HttpProtocols = listenOptions.Protocols;
```

Reply:
The first thing `ApplyHttpsDefaults`, is a thing we want to delay, and this check has to come after that. I think this delays the error from when the host resolves the IServer until when it starts it. We should make sure exceptions still cause app.Run() to throw at startup, but I think having IServer.StartAsync throw is a fine way to achieve this.

Reply:
I just want to make sure we know what the new failure mode looks like.

Reply:
Old
```
>	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Hosting.ListenOptionsHttpsExtensions.UseHttps(Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions listenOptions, System.Action<Microsoft.AspNetCore.Server.Kestrel.Https.HttpsConnectionAdapterOptions> configureOptions) Line 186	C#
 	WebApplication51.dll!Program.<Main>$.AnonymousMethod__0_1(Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions listenOptions) Line 11	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions.Listen(System.Net.EndPoint endPoint, System.Action<Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions> configure) Line 435	C#
 	WebApplication51.dll!Program.<Main>$.AnonymousMethod__0_0(Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions options) Line 10	C#
 	Microsoft.Extensions.Options.dll!Microsoft.Extensions.Options.OptionsFactory<Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions>.Create(string name)	Unknown
 	Microsoft.Extensions.Options.dll!Microsoft.Extensions.Options.UnnamedOptionsManager<Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions>.Value.get()	Unknown
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.CreateServiceContext(Microsoft.Extensions.Options.IOptions<Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions> options, Microsoft.Extensions.Logging.ILoggerFactory loggerFactory, System.Diagnostics.DiagnosticSource diagnosticSource) Line 110	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.KestrelServerImpl(Microsoft.Extensions.Options.IOptions<Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerOptions> options, System.Collections.Generic.IEnumerable<Microsoft.AspNetCore.Connections.IConnectionListenerFactory> transportFactories, System.Collections.Generic.IEnumerable<Microsoft.AspNetCore.Connections.IMultiplexedConnectionListenerFactory> multiplexedFactories, Microsoft.Extensions.Logging.ILoggerFactory loggerFactory, System.Diagnostics.DiagnosticSource diagnosticSource) Line 59	C#
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(Microsoft.Extensions.DependencyInjection.ServiceLookup.ConstructorCallSite constructorCallSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, System.__Canon>.VisitCallSiteMain(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext argument)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext context)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor<Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, object>.VisitCallSite(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext argument)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceCallSite callSite, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope scope)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(System.Type serviceType)	Unknown
 	System.Collections.Concurrent.dll!System.Collections.Concurrent.ConcurrentDictionary<System.Type, System.Func<Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope, object>>.GetOrAdd(System.Type key, System.Func<System.Type, System.Func<Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope, object>> valueFactory)	Unknown
 	Microsoft.Extensions.DependencyInjection.dll!Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(System.Type serviceType, Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope serviceProviderEngineScope)	Unknown
 	Microsoft.Extensions.DependencyInjection.Abstractions.dll!Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(System.IServiceProvider provider, System.Type serviceType)	Unknown
 	Microsoft.Extensions.DependencyInjection.Abstractions.dll!Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService<Microsoft.AspNetCore.Hosting.Server.IServer>(System.IServiceProvider provider)	Unknown
 	Microsoft.AspNetCore.dll!Microsoft.AspNetCore.Builder.WebApplication.ServerFeatures.get()	Unknown
 	Microsoft.AspNetCore.dll!Microsoft.AspNetCore.Builder.WebApplication.WebApplication(Microsoft.Extensions.Hosting.IHost host)	Unknown
 	Microsoft.AspNetCore.dll!Microsoft.AspNetCore.Builder.WebApplicationBuilder.Build()	Unknown
 	WebApplication51.dll!Program.<Main>$(string[] args) Line 15	C#
```

New
```
>	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Hosting.ListenOptionsHttpsExtensions.UseHttps.AnonymousMethod__0() Line 180	C#
 	System.Private.CoreLib.dll!System.Lazy<Microsoft.AspNetCore.Server.Kestrel.Https.HttpsConnectionAdapterOptions>.ViaFactory(System.Threading.LazyThreadSafetyMode mode)	Unknown
 	System.Private.CoreLib.dll!System.Lazy<Microsoft.AspNetCore.Server.Kestrel.Https.HttpsConnectionAdapterOptions>.ExecutionAndPublication(System.LazyHelper executionAndPublication, bool useDefaultConstructor)	Unknown
 	System.Private.CoreLib.dll!System.Lazy<Microsoft.AspNetCore.Server.Kestrel.Https.HttpsConnectionAdapterOptions>.CreateValue()	Unknown
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Hosting.ListenOptionsHttpsExtensions.UseHttps.AnonymousMethod__0(Microsoft.AspNetCore.Connections.ConnectionDelegate next) Line 219	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions.Build() Line 201	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.StartAsync.__OnBind|0(Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions options, System.Threading.CancellationToken onBindCancellationToken) Line 206	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions endpoint, Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBindContext context, System.Threading.CancellationToken cancellationToken) Line 89	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions.BindAsync(Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBindContext context, System.Threading.CancellationToken cancellationToken) Line 225	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.EndpointsStrategy.BindAsync(Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBindContext context, System.Threading.CancellationToken cancellationToken) Line 210	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.OverrideWithEndpointsStrategy.BindAsync(Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBindContext context, System.Threading.CancellationToken cancellationToken) Line 193	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindAsync(System.Collections.Generic.IEnumerable<Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions> listenOptions, Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBindContext context, System.Threading.CancellationToken cancellationToken) Line 31	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.BindAsync(System.Threading.CancellationToken cancellationToken) Line 317	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.StartAsync<Microsoft.AspNetCore.Hosting.HostingApplication.Context>(Microsoft.AspNetCore.Hosting.Server.IHttpApplication<Microsoft.AspNetCore.Hosting.HostingApplication.Context> application, System.Threading.CancellationToken cancellationToken) Line 237	C#
 	Microsoft.AspNetCore.Hosting.dll!Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(System.Threading.CancellationToken cancellationToken) Line 158	C#
 	Microsoft.Extensions.Hosting.dll!Microsoft.Extensions.Hosting.Internal.Host.StartAsync(System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.Extensions.Hosting.Abstractions.dll!Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(Microsoft.Extensions.Hosting.IHost host, System.Threading.CancellationToken token)	Unknown
 	PlaintextApp.dll!PlaintextApp.Startup.Main(string[] args) Line 59	C#
 	PlaintextApp.dll!PlaintextApp.Startup.<Main>(string[] args)	Unknown
```

Reply:
AFAIK, we _can't_ check as early as we did because certs from config won't be loaded yet.

Reply:
I'm taking thumbs-up to mean we're satisfied with the new experience.   It's a bit worse, but seems worthwhile to be able to pick up `IConfiguration` consistently.